essential-projects/metadata
# 🚸 add global-variable-metadata-provider-hack

## What did you change?

*The PR shows that about every line was changed, but thats only because they are now all indented one level further. The actual changes aren't big!*

The MetadataProvicer is currently the only implementation in the stack that
cannot be provided via ioc, because it is used in decorator-functions. The
reason being that the decorator-functions are executed the moment a class-file
is `require`d, wich is way to early for dependency-injection.

This is not a problem in usual setups like the meta setup or a regular
dependency-install from npm, as you'd usually only have one instance of the
`metadata`-module in the node_modules.

If for some reason you were to have multiple instances of the `metadata` package
somewhere in the node_modules-dependency-folder-structure you have the risk that
not all metadata (like schemas) is registered to the correct MetadataProvider.

This happened to me when i tried to test changes to one package using our
integration-tests. I ran `npm install` and `npm test`, and everything was fine.
Then i wanted to test changes in the `process_engine`-package, so i ran
`npm link @proces-engine/process_engine`. Because the linked
process_engine-package had its own node_modules including the `metadata`
package, all entity-decorators used the MetadataProvider from that package,
instead of the `metadata`-package the tests themselves used.

This PR is not a fix, but a workaround. It makes the the Metadata register
itself to a global variable. When it detects that another MetadataProvider
already registered itself to that variable, it logs a warning and exports that
already existing MetadataProvider instead of a new one.

## How can others test the changes?

1. prepare a fresh postgres-database as described in the [skeleton-project](https://github.com/process-engine/skeleton/tree/develop/database)
2. clone [process_engine_meta](https://github.com/process-engine/process_engine_meta/tree/feature/add_integration_tests) on `feature/add_integration_tests`
3. cd into `_integration_tests`
4. run `npm install`
5. run `npm test`
6. **see how the tests work**
7. in a different folder, clone [process_engine](https://github.com/process-engine/process_engine) on `develop`
8. run `npm link` in the `process_engine` folder
9. cd back into the `_integration_tests` folder
10. run `npm link @process-engine/process_engine`
11. run `npm test`
12. **see how the tests no longer work**
13. in a different folder clone metadata on
    `feature/meta_multi_registration_workaround`
14. cd into the `metadata` folder and run `npm install` and `npm run build`
14. copy the contents from `metadata/dist/commonjs/provider.js` to
    `_integration_tests/node_modules/@essential-projects/metadata/dist/commonjs/provider.js`
    **AND** to
    `_integration_tests/node_modules/@process-engine/process_engine/node_modules/@essential-projects/metadata/dist/commonjs/provider.js`
15. run `npm test` again
16. **see how the tests work again** (and how there is now a warning log because of the multiple metadata-packages)


## PR-Checklist

Please check the boxes in this list after submitting your PR:
